File: 필수문항/135_물병.py
import sys

input = sys.stdin.readline

N, K = map(int, input().split())

bi_N = list(bin(N)[2:][::-1])
now = bi_N.count("1")
ans = 0

while now > K:
    for i in range(len(bi_N)):
        if bi_N[i] == "1":
            ans += 2**i
            bi_N[i] = "0"
            tmp = i
            while tmp < len(bi_N) - 1 and bi_N[tmp + 1] == "1":
                now -= 1
                bi_N[tmp + 1] = "0"
                tmp += 1

            if tmp == len(bi_N) - 1:
                bi_N += "1"
            else:
                bi_N[tmp + 1] = "1"
            break

print(ans)

File: 필수문항/135_물병.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import java.util.StringTokenizer;

public class Main {

    public static void main(String[] args) throws IOException {
        int ans = 0;
        int now = 0;

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        String num = Integer.toBinaryString(Integer.parseInt(st.nextToken()));
        List<Integer> N = new ArrayList<>();
        for (int i = num.length() - 1; i >= 0; i--) {
            int tmp = Character.getNumericValue(num.charAt(i));
            N.add(tmp);
            if (tmp == 1) {
                now++;
            }
        }

        int K = Integer.parseInt(st.nextToken());

        while (now > K) {
            for (int i = 0; i < N.size(); i++) {
                if (N.get(i) == 1) {
                    ans += Math.pow(2, i);
                    N.set(i, 0);
                    int tmp = i;
                    while (tmp < N.size() - 1 && N.get(tmp + 1) == 1) {
                        tmp++;
                        now--;
                        N.set(tmp, 0);
                    }
                    if (tmp == N.size() - 1) {
                        N.add(1);
                    } else {
                        N.set(tmp + 1, 1);
                    }
                    break;
                }
            }
        }
        System.out.println(ans);
    }
}
